sharktank/sharktank/ops/sharded_impls.py

```diff
@@ -1407,6 +1407,85 @@ def reshard_like_unreduced_to_replicated(
     return replicate(tensor, count=like.shard_count)
 
 
+@scatter_.override(ReplicatedTensor, ReplicatedTensor)
+def scatter_replicated_replicated(
+    inout: ReplicatedTensor,
+    dim: int,
+    index: ReplicatedTensor,
+    value,
+    *,
+    reduce: str = None,
+) -> ReplicatedTensor:
+    assert inout.shard_count == index.shard_count
+    for shard, index_shard in zip(inout.shards, index.shards):
+        scatter_(shard, dim, index_shard, value, reduce=reduce)
+    return inout
+
+
+@scatter_.override(SplitPrimitiveTensor, SplitPrimitiveTensor)
+def scatter_split_split(
+    inout: SplitPrimitiveTensor,
+    dim: int,
+    index: SplitPrimitiveTensor,
+    value,
+    *,
+    reduce: str = None,
+) -> SplitPrimitiveTensor:
+    if dim == inout.shard_dim:
+        # `index` can contain indices into any of `inout`s shards in any of its entries.
+        # Can't know ahead of time how to seperate out its values based on sliices.
+        tmp_tensor = all_gather(inout)
+        index = all_gather(index)
+        tmp_tensor.scatter_(dim, index, value, reduce=reduce)
+        tmp_tensor = reshard_like(tmp_tensor, inout)
+
+        for inout_shard, tmp_shard in zip(inout.shards, tmp_tensor.shards):
+            inout_shard.as_torch().copy_(tmp_shard.as_torch())
+        return inout
+
+    shard_dim = inout.shard_dim
+    if index.shape[shard_dim] == inout.shape[shard_dim]:
+        assert index.shard_dim == inout.shard_dim
+        index_shards = index.shards
+        last_shard_idx = inout.shard_count - 1
+    else:
```

Comment:
I don't understand that case.
After all-gathering the `index`, shouldn't we reshard it like the `inout`?
Then we can scatter locally for each pair of corresponding `inout` and index shards.

Reply:
By default we are sharding both `inout` and `index` to be balanced. However, if `index` has fewer entries along the shard_dim that `inout` then entry `i` along `shard_dim` that's in shard `j` of `inout` may end up in shard `k` of index. We need to reshard so that slices along `shard_dim` within each shard line up.